### ISSUE TITLE
io/console.rb on AIX not working in v9.x

### ISSUE THREAD

**Issue by ssteidte** [jruby/jruby#3495]
On AIX, io/console.rb is broken for all versions of JRuby 9k including 9.0.4.
The error displayed is

`
TypeError: no implicit conversion of nil into String
   reopen at org/jruby/RubyIO.java:582
    <top> at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/io/console.rb:202
`

The reason is that the call $stderr.dup in line 195 fails:
`NotImplementedError: dup unsupported or native support failed to load`

This error is caught, but the subsequent try to reopen old_stderr fails in line 202.

**Comment by tajila** [comment]
I am seeing the same issue on AIX (64bit JVM) using JRuby v9.0.4. Here is the full stack trace:

	TypeError: no implicit conversion of nil into String
	   reopen at org/jruby/RubyIO.java:582
		<top> at /team/tobi/PR106886/InvokeDynamicRubyv904_Package/InvokeDynamic_Ruby_v904/lib/ruby/stdlib/io/console.rb:202
	  require at org/jruby/RubyKernel.java:939
		<top> at /team/tobi/PR106886/InvokeDynamicRubyv904_Package/InvokeDynamic_Ruby_v904/lib/ruby/stdlib/rubygems/user_interaction.rb:1
	  require at org/jruby/RubyKernel.java:939
		<top> at /team/tobi/PR106886/InvokeDynamicRubyv904_Package/InvokeDynamic_Ruby_v904/lib/ruby/stdlib/rubygems/user_interaction.rb:8
	  require at org/jruby/RubyKernel.java:939
		<top> at /team/tobi/PR106886/InvokeDynamicRubyv904_Package/InvokeDynamic_Ruby_v904/lib/ruby/stdlib/rubygems/config_file.rb:1
	  require at org/jruby/RubyKernel.java:939
		<top> at /team/tobi/PR106886/InvokeDynamicRubyv904_Package/InvokeDynamic_Ruby_v904/lib/ruby/stdlib/rubygems/config_file.rb:7
	  require at org/jruby/RubyKernel.java:939
		<top> at /team/tobi/PR106886/InvokeDynamicRubyv904_Package/InvokeDynamic_Ruby_v904/lib/ruby/stdlib/rubygems/defaults/jruby.rb:1
		 load at org/jruby/RubyKernel.java:957
		<top> at /team/tobi/PR106886/InvokeDynamicRubyv904_Package/InvokeDynamic_Ruby_v904/lib/ruby/stdlib/rubygems/defaults/jruby.rb:1
	
Just as @ssteidte stated the cause of the failure is line 195 in io/console.rb:

	old_stderr = $stderr.dup

$stderr.dup throws an error (NotImplementedError: dup unsupported or native support failed to load)

In the ensure block (line 202) the call $stderr.reopen(old_stderr) fails because old_stderr is nil, it was never set.

I do not see this issue when using a 32bit JVM on AIX.

**Comment by headius** [comment]
Can one of you show me the result of booting JRuby with `-Xnative.logging=true` please?

There's a couple possibilities:

* We may not be finding the right libraries to load in 64-bit mode
* The functions we need may need to be bound differently on AIX 64 bit.

Please also provide environment information, like `jruby --version` and `uname -a`.

**Comment by ssteidte** [comment]
the switch doesn't work:
`java -cp jruby-complete-9.0.5.0.jar org.jruby.Main -Xnative.logging=true` gives
`jruby: invalid extended option native.logging=true (-X will list valid options)`

jruby -version:
`jruby 9.0.5.0 (2.2.3) 2016-01-26 7bee00d IBM J9 VM 2.7 on pap6470_27sr3fp10-20150708_01 (SR3 FP10) +jit [AIX-ppc64]`

uname -a: `AIX f36smd1 1 7 00F646FB4C00`

**Comment by ssteidte** [comment]
sorry, didn't want to close it

**Comment by headius** [comment]
@ssteidte Sorry, if passing flags to `java` you'd want the long form: `-Djruby.native.logging=true`.

**Comment by ssteidte** [comment]
still doesn't recognize the switch: `jruby: unknown option -Djruby.native.logging=true` with JRuby 9.0.5

**Comment by headius** [comment]
`java -Djruby.native.logging=true -cp jruby-complete-9.0.5.0.jar org.jruby.Main ...`

**Comment by ssteidte** [comment]
>  java7/jre/bin/java -Djruby.native.logging=true -cp jruby-complete-9.0.5.0.jar org.jruby.Main
jruby: warning: unknown property jruby.native.logging

**Comment by enebo** [comment]
@ssteidte -Djruby.native.verbose=false

**Comment by headius** [comment]
Oops, my mistake...@enebo is right, it's `-Djruby.native.verbose=<true|false>`.

**Comment by ssteidte** [comment]
now this works :)

Failed to load native POSIX impl; falling back on Java impl. Stacktrace follows.
java.lang.UnsatisfiedLinkError: could not load FFI provider jnr.ffi.provider.jffi.Provider
        at jnr.ffi.provider.InvalidRuntime.newLoadError(InvalidRuntime.java:101)
        at jnr.ffi.provider.InvalidRuntime.addressSize(InvalidRuntime.java:82)
        at jnr.posix.POSIXFactory.libraries(POSIXFactory.java:183)
        at jnr.posix.POSIXFactory.access$200(POSIXFactory.java:16)
        at jnr.posix.POSIXFactory$DefaultLibCProvider$SingletonHolder.<clinit>(POSIXFactory.java:279)
        at jnr.posix.POSIXFactory$DefaultLibCProvider.getLibC(POSIXFactory.java:283)
        at jnr.posix.BaseNativePOSIX.<init>(BaseNativePOSIX.java:36)
        at jnr.posix.AixPOSIX.<init>(AixPOSIX.java:43)
        at jnr.posix.POSIXFactory.loadAixPOSIX(POSIXFactory.java:167)
        at jnr.posix.POSIXFactory.loadNativePOSIX(POSIXFactory.java:133)
        at jnr.posix.POSIXFactory.loadPOSIX(POSIXFactory.java:90)
        at jnr.posix.LazyPOSIX.loadPOSIX(LazyPOSIX.java:36)
        at jnr.posix.LazyPOSIX.posix(LazyPOSIX.java:32)
        at jnr.posix.LazyPOSIX.isNative(LazyPOSIX.java:370)
        at org.jruby.RubyGlobal.prepareStdioChannel(RubyGlobal.java:279)
        at org.jruby.RubyGlobal.createGlobals(RubyGlobal.java:193)
        at org.jruby.Ruby.init(Ruby.java:1241)
        at org.jruby.Ruby.newInstance(Ruby.java:330)
        at org.jruby.Main.internalRun(Main.java:278)
        at org.jruby.Main.run(Main.java:239)
        at org.jruby.Main.main(Main.java:201)
Caused by: java.lang.ExceptionInInitializerError
        at java.lang.J9VMInternals.ensureError(J9VMInternals.java:134)
        at java.lang.J9VMInternals.recordInitializationFailure(J9VMInternals.java:123)
        at jnr.ffi.provider.jffi.NativeRuntime.getInstance(NativeRuntime.java:49)
        at jnr.ffi.provider.jffi.Provider.<init>(Provider.java:29)
        at java.lang.J9VMInternals.newInstanceImpl(Native Method)
        at java.lang.Class.newInstance(Class.java:1920)
        at jnr.ffi.provider.FFIProvider$SystemProviderSingletonHolder.getInstance(FFIProvider.java:68)
        at jnr.ffi.provider.FFIProvider$SystemProviderSingletonHolder.<clinit>(FFIProvider.java:57)
        at jnr.ffi.provider.FFIProvider.getSystemProvider(FFIProvider.java:35)
        at jnr.ffi.Runtime$SingletonHolder.<clinit>(Runtime.java:82)
        at jnr.ffi.Runtime.getSystemRuntime(Runtime.java:67)
        ... 19 more
Caused by: java.lang.IllegalStateException: Cause already initialized
        at java.lang.Throwable.setCause(Throwable.java:361)
        at java.lang.Throwable.initCause(Throwable.java:339)
        at com.kenai.jffi.Type$Builtin.lookupTypeInfo(Type.java:252)
        at com.kenai.jffi.Type$Builtin.getTypeInfo(Type.java:237)
        at com.kenai.jffi.Type.resolveSize(Type.java:155)
        at com.kenai.jffi.Type.size(Type.java:138)
        at jnr.ffi.provider.jffi.NativeRuntime$TypeDelegate.size(NativeRuntime.java:178)
        at jnr.ffi.provider.AbstractRuntime.<init>(AbstractRuntime.java:48)
        at jnr.ffi.provider.jffi.NativeRuntime.<init>(NativeRuntime.java:57)
        at jnr.ffi.provider.jffi.NativeRuntime.<init>(NativeRuntime.java:41)
        at jnr.ffi.provider.jffi.NativeRuntime$SingletonHolder.<clinit>(NativeRuntime.java:53)
        ... 28 more
TypeError: no implicit conversion of nil into String
   reopen at org/jruby/RubyIO.java:582

**Comment by ssteidte** [comment]
uname -a: AIX f36smd1 1 7 00F646FB4C00

java --version:
>  java7/jre/bin/java -version
java version "1.7.0"
Java(TM) SE Runtime Environment (build pap6470_27sr3fp10-20150708_01(SR3 FP10))
IBM J9 VM (build 2.7, JRE 1.7.0 AIX ppc64-64 Compressed References 20150630_255653 (JIT enabled, AOT enabled)
J9VM - R27_Java727_SR3_20150630_2236_B255653
JIT  - tr.r13.java_20150623_94888.01
GC   - R27_Java727_SR3_20150630_2236_B255653_CMPRSS
J9CL - 20150630_255653)
JCL - 20150628_01 based on Oracle jdk7u85-b15

**Comment by headius** [comment]
@ssteidte Is that the complete output from `native.verbose`? If so, it looks like this might be a simple bug in JFFI.

I don't suppose there's any way to get shell access to an equivalent environment, so we can debug and test this in place?

In the interim, my recommendation to you (if you have cycles to work with us) would be to start checking out JNR projects and see if they build and test successfully, starting with https://github.com/jnr/jffi and then moving on to jnr-ffi and jnr-posix. I suspect at least one of them will fail to run, which will give us something concrete to work on.

**Comment by headius** [comment]
FYI, all the JNR projects are straightforward maven projects; `mvn package` should be sufficient to build and run tests.

**Comment by ssteidte** [comment]
Yes, that was the whole output (I've just leaved out the Ruby part that is shown in the first post).
Unfortunately I don't have enough space on the AIX machine to do substantial stuff like compilations.
Anything else, like test runs - volontarily.

**Comment by headius** [comment]
Ok, well I'll stick with you until we can figure this out. I'm poking around for AIX cloud options now.

Meanwhile...

I think the general situation here is that our native support for AIX is failing to load, which in turn breaks a lot of things that expect to use native calls on *nix platforms. So he first thing to do is to figure out why it fails to load.

I need from you:

* uname -a (or as much of it as you're comfortable revealing)
* JDK version
* Output of `jruby -rpp -e "pp RbConfig::CONFIG"`

If it's at all possible to try building jffi on AIX, that's going to be our fastest path forward. If you can stop by our IRC chat room today (http://chat.jruby.org) that would also help things along.

**Comment by headius** [comment]
If you decide you want to try the jffi build (requires Apache Maven, any recent version):

```
git clone https://github.com/jnr/jffi
cd jffi
mvn clean test
```

...and we can iterate from there.

**Comment by ssteidte** [comment]
uname -a:
AIX f36smd1 1 7 00F646FB4C00

java -version:
java version "1.7.0"
Java(TM) SE Runtime Environment (build pap6470_27sr3fp10-20150708_01(SR3 FP10))
IBM J9 VM (build 2.7, JRE 1.7.0 AIX ppc64-64 Compressed References 20150630_255653 (JIT enabled, AOT enabled)
J9VM - R27_Java727_SR3_20150630_2236_B255653
JIT  - tr.r13.java_20150623_94888.01
GC   - R27_Java727_SR3_20150630_2236_B255653_CMPRSS
J9CL - 20150630_255653)
JCL - 20150628_01 based on Oracle jdk7u85-b15

jruby -rpp -e "pp RbConfig::CONFIG": gives the error that this is all about

**Comment by headius** [comment]
Anyone manage to get an AIX build of jffi working? Going to have to punt this since we don't really have a way to investigate it. We need your help.

Note that fixes in jnr/jnr-enxio#12 may have improved this bug too, and it should be shipping with latest JRuby versions (1.7.25 and 9.1).

**Comment by tajila** [comment]
I did the following:
`git clone https://github.com/jnr/jffi`
`cd jffi`
`mvn clean test`

and got the following output:

[INFO] Scanning for projects...
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building jffi 1.2.12-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ jffi ---
[INFO] Deleting /team/tobi/jffibug/jffi/target
[INFO]
[INFO] --- maven-enforcer-plugin:1.0:enforce (enforce-maven) @ jffi ---
[INFO]
[INFO] --- maven-antrun-plugin:1.6:run (build-native-library) @ jffi ---
[WARNING] Parameter tasks is deprecated, use target instead
[INFO] Executing tasks

main:

-generate-version-source:
     [echo] Generating Version.java
    [mkdir] Created dir: /team/tobi/jffibug/jffi/target/classes
    [mkdir] Created dir: /team/tobi/jffibug/jffi/target/java/com/kenai/jffi
[INFO] Executed tasks
[INFO] Registering compile source root /team/tobi/jffibug/jffi/target/java
[INFO]
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ jffi ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /team/tobi/jffibug/jffi/src/main/resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ jffi ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 43 source files to /team/tobi/jffibug/jffi/target/classes
[WARNING] bootstrap class path not set in conjunction with -source 1.6
[WARNING] /team/tobi/jffibug/jffi/src/main/java/com/kenai/jffi/ClosureMagazine.java:[57,21] free() in com.kenai.jffi.Closure.Handle has been deprecated
[WARNING] /team/tobi/jffibug/jffi/src/main/java/com/kenai/jffi/HeapInvocationBuffer.java: Some input files use unchecked or unsafe operations.
[WARNING] /team/tobi/jffibug/jffi/src/main/java/com/kenai/jffi/HeapInvocationBuffer.java: Recompile with -Xlint:unchecked for details.
[INFO]
[INFO] --- maven-bundle-plugin:2.4.0:manifest (jffi-complete) @ jffi ---
[INFO]
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ jffi ---
[INFO] Not copying test resources
[INFO]
[INFO] --- maven-antrun-plugin:1.6:run (default) @ jffi ---
[WARNING] Parameter tasks is deprecated, use target instead
[INFO] Executing tasks

main:
    [unzip] Expanding: /team/tobi/jffibug/jffi/archive/jffi-Darwin.jar into /team/tobi/jffibug/jffi/target
    [unzip] Expanding: /team/tobi/jffibug/jffi/archive/jffi-arm-Linux.jar into /team/tobi/jffibug/jffi/target
    [unzip] Expanding: /team/tobi/jffibug/jffi/archive/jffi-i386-FreeBSD.jar into /team/tobi/jffibug/jffi/target
    [unzip] Expanding: /team/tobi/jffibug/jffi/archive/jffi-i386-Linux.jar into /team/tobi/jffibug/jffi/target
    [unzip] Expanding: /team/tobi/jffibug/jffi/archive/jffi-i386-OpenBSD.jar into /team/tobi/jffibug/jffi/target
    [unzip] Expanding: /team/tobi/jffibug/jffi/archive/jffi-i386-SunOS.jar into /team/tobi/jffibug/jffi/target
    [unzip] Expanding: /team/tobi/jffibug/jffi/archive/jffi-i386-Windows.jar into /team/tobi/jffibug/jffi/target
    [unzip] Expanding: /team/tobi/jffibug/jffi/archive/jffi-ppc-AIX.jar into /team/tobi/jffibug/jffi/target
    [unzip] Expanding: /team/tobi/jffibug/jffi/archive/jffi-ppc-Linux.jar into /team/tobi/jffibug/jffi/target
    [unzip] Expanding: /team/tobi/jffibug/jffi/archive/jffi-ppc64-Linux.jar into /team/tobi/jffibug/jffi/target
    [unzip] Expanding: /team/tobi/jffibug/jffi/archive/jffi-ppc64le-Linux.jar into /team/tobi/jffibug/jffi/target
    [unzip] Expanding: /team/tobi/jffibug/jffi/archive/jffi-s390x-Linux.jar into /team/tobi/jffibug/jffi/target
    [unzip] Expanding: /team/tobi/jffibug/jffi/archive/jffi-sparc-SunOS.jar into /team/tobi/jffibug/jffi/target
    [unzip] Expanding: /team/tobi/jffibug/jffi/archive/jffi-sparcv9-SunOS.jar into /team/tobi/jffibug/jffi/target
    [unzip] Expanding: /team/tobi/jffibug/jffi/archive/jffi-x86_64-FreeBSD.jar into /team/tobi/jffibug/jffi/target
    [unzip] Expanding: /team/tobi/jffibug/jffi/archive/jffi-x86_64-Linux.jar into /team/tobi/jffibug/jffi/target
    [unzip] Expanding: /team/tobi/jffibug/jffi/archive/jffi-x86_64-OpenBSD.jar into /team/tobi/jffibug/jffi/target
    [unzip] Expanding: /team/tobi/jffibug/jffi/archive/jffi-x86_64-SunOS.jar into /team/tobi/jffibug/jffi/target
    [unzip] Expanding: /team/tobi/jffibug/jffi/archive/jffi-x86_64-Windows.jar into /team/tobi/jffibug/jffi/target
[INFO] Executed tasks
[INFO]
[INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ jffi ---
[INFO] Not compiling test sources
[INFO]
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ jffi ---
[INFO] Tests are skipped.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 25.611 s
[INFO] Finished at: 2016-04-25T14:42:07-04:00
[INFO] Final Memory: 20M/34M
[INFO] ------------------------------------------------------------------------

**Comment by tajila** [comment]
I wrote a small test program
`jnr.ffi.Runtime rt = FFIProvider.getSystemProvider().getRuntime();`
`System.out.println("rt: " + rt);`
`System.out.println("getAddressSize " + rt.addressSize());`

on with a 64-bit jvm the result was:
rt: jnr.ffi.provider.InvalidRuntime@e2f51c1a
Exception in thread "main" java.lang.UnsatisfiedLinkError: could not load FFI provider jnr.ffi.provider.jffi.Provider
        at jnr.ffi.provider.InvalidRuntime.newLoadError(InvalidRuntime.java:101)
        at jnr.ffi.provider.InvalidRuntime.addressSize(InvalidRuntime.java:82)
        at Test.main(Test.java:11)
Caused by: java.lang.ExceptionInInitializerError
        at java.lang.J9VMInternals.ensureError(J9VMInternals.java:134)
        at java.lang.J9VMInternals.recordInitializationFailure(J9VMInternals.java:123)
        at jnr.ffi.provider.jffi.NativeRuntime.getInstance(NativeRuntime.java:49)
        at jnr.ffi.provider.jffi.Provider.<init>(Provider.java:29)
        at java.lang.J9VMInternals.newInstanceImpl(Native Method)
        at java.lang.Class.newInstance(Class.java:1911)
        at jnr.ffi.provider.FFIProvider$SystemProviderSingletonHolder.getInstance(FFIProvider.java:68)
        at jnr.ffi.provider.FFIProvider$SystemProviderSingletonHolder.<clinit>(FFIProvider.java:57)
        at jnr.ffi.provider.FFIProvider.getSystemProvider(FFIProvider.java:35)
        at Test.main(Test.java:9)
Caused by: java.lang.IllegalStateException: Cause already initialized
        at java.lang.Throwable.setCause(Throwable.java:365)
        at java.lang.Throwable.initCause(Throwable.java:341)
        at com.kenai.jffi.Type$Builtin.lookupTypeInfo(Type.java:252)
        at com.kenai.jffi.Type$Builtin.getTypeInfo(Type.java:237)
        at com.kenai.jffi.Type.resolveSize(Type.java:155)
        at com.kenai.jffi.Type.size(Type.java:138)
        at jnr.ffi.provider.jffi.NativeRuntime$TypeDelegate.size(NativeRuntime.java:178)
        at jnr.ffi.provider.AbstractRuntime.<init>(AbstractRuntime.java:48)
        at jnr.ffi.provider.jffi.NativeRuntime.<init>(NativeRuntime.java:57)
        at jnr.ffi.provider.jffi.NativeRuntime.<init>(NativeRuntime.java:41)
        at jnr.ffi.provider.jffi.NativeRuntime$SingletonHolder.<clinit>(NativeRuntime.java:53)
        ... 8 more

on 32 bit the result was:
rt: jnr.ffi.provider.jffi.NativeRuntime@1a160cbb
getAddressSize 4

So I think this little experiment proves it's only an issue for 64bit JVM's on AIX.

After some digging I found that the failure is caused by doing a System.load(...) on /tmp/jffi394152621522646037.so in StubLoader.Load(). The error is:
java.lang.UnsatisfiedLinkError: /tmp/jffi8307107834842217746.so (	0509-022 Cannot load module /tmp/jffi8307107834842217746.so.
	0509-026 System error: Cannot run a file that does not have a valid format.)

I added the directory of jffi-1.2.a to the "-Djava.library.path" so it would run StubLoader.loadFromBootPath(...) instead of  StubLoader.loadFromJar(...) but the results were the same.
(0509-022 Cannot load module ..snip../lib/libjffi-1.2.a.
	0509-026 System error: Cannot run a file that does not have a valid format.)

Note: I had to fiddle with StubLoader.getPlatformName() to get as far as I did because  StubLoader.getStubLibraryPath returned `jni/ppc64-AIX/...` which doesn't exist. I modified it to return `jni/ppc-AIX/...`

After I tried:
`nm -X64 target/jni/ppc-AIX/libjffi-1.2.a`
0654-210 target/jni/ppc-AIX/libjffi-1.2.a is not valid in the current object file mode.
        Use the -X option to specify the desired object mode.

but `nm -X32 target/jni/ppc-AIX/libjffi-1.2.a`
nm: target/jni/ppc-AIX/libjffi-1.2.a: 0654-204 There are no symbols in the file.

So my guess is that "libjffi-1.2.a" is not built for a 64bit environment. Or, jruby.jar is missing the 64 bit version of libjffi-1.2.a

**Comment by headius** [comment]
@tajila Sorry that we have not revisited this in a long time. It sounds like the whole problem lies within jffi. Can you provide us a proper AIX 64-bit build of `libjffi-1.2.a`, and perhaps help us patch the build to produce it correctly?

**Comment by headius** [comment]
Punting to 9.1.4.0. Hopefully @tajila can hook us up with a good jffi for 64-bit AIX :-)

**Comment by tajila** [comment]
I'll look into it. Maybe in the next week or so I can give you an estimate for completion.

**Comment by ssteidte** [comment]
Assuming jffi as culprit seems to be right. Tried the brand new version 9.1.3 and the error is now much later after the startup:
NotImplementedError: flock unsupported or native support failed to load; see http://wiki.jruby.org/Native-Libraries
           flock at org/jruby/RubyFile.java:318
  create_logfile at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/logger.rb:712

**Comment by headius** [comment]
If the error is much later than startup, I start to wonder if it's fixed but something unrelated is now broken. We have had some issues with flock not booting on platforms where fcntl is preferred for file-locking (like several non-Linux, non-BSD OSes) and have just added the function correctly or worked to reimplement it using appropriate calls (see jnr/jnr-posix#60, for example).

So, if we can patch around this we may find things are working correctly.

**Comment by headius** [comment]
Link with #4112.

**Comment by ssteidte** [comment]
update for JRuby 9.1.7: basically the same as in 9.1.3

NotImplementedError: flock unsupported or native support failed to load; see http://wiki.jruby.org/Native-Libraries
           flock at org/jruby/RubyFile.java:324
  create_logfile at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/logger.rb:712
    open_logfile at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/logger.rb:705
         set_dev at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/logger.rb:695
      initialize at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/logger.rb:635
      initialize at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/logger.rb:353

**Comment by headius** [comment]
Damn flock again!

So is there no flock on AIX? It sounds like the FFI library is loading but it can't bind some things. This would be a situation similar to Solaris, where we must use fcntl to emulate flock.

**Comment by iconoclast** [comment]
I seem to be running into this bug as well. The original "TypeError: no implicit conversion of nil into String", that is, I haven't made it as far as the flock issue yet.

The comments above seem to suggest that an update to libjffi got past the original error message, but I'm still running into that error with JRuby 9.1.7.0.  Is there some piece I'm missing that doesn't involve compiling jffi and friends from scratch?

**Comment by iconoclast** [comment]
Although it's true that libjffi-1.2.a is broken on 64bit AIX and should be certainly be fixed, that may be tangential to the issue here.  It turns out that libjffi was borked in JRuby 1.7 as well, yet the code I have triggering this error on 9k runs fine on 1.7.

A simple require "io/console" on 1.7 will give a warning of "io/console not supported; tty will not be manipulated" but then the program continues to run.
The same require statement on 9.1.7.0 will die with a stack trace like the one at the top of this ticket.
In both cases jruby.native.verbose shows the "Failed to load native POSIX impl..." message and stacktrace like the one seen earlier in this ticket.

The relevant portion of console.rb is essentially the same in both versions of JRuby.  The error shown at the start of this ticket is being triggered by $stderr.reopen but only because there is a rescue clause masking the previous error triggered by $stderr.dup a few lines earlier.  Both of these methods appear to work in 1.7.  Since this code is outside of any class definition it is run as soon as the require occurs which also makes it difficult to monkeypatch around it.

Digging further down the rabbit hole takes us into RubyIO.java which is significantly different in 9k.  At a glance I didn't spot any obvious reason why 1.7 has a working fallback and 9k doesn't, but I'm also probably in over my head at this point.

So although a working libjffi would be ideal, it seems like JRuby 1.7 had enough of a fallback implementation to continue to work well enough (for some anyway) even in the absence of a working libjffi, while 9k does not.

**Comment by headius** [comment]
@iconoclast What are you running that produces the type error?

The methods you say work fine in 1.7 are likely not dependent on native operations at all, preferring Java/JDK equivalents. So there's some platform exercise needed.

It would be a lot easier to maintain AIX support if it were possible for us to run it locally and in CI :cry: I'll poke around and see if it's possible to do that today.

**Comment by iconoclast** [comment]
@headius Reproducing this is as simple as requiring 'io/console'.

```
bash-3.2$ java -jar jruby-complete-1.7.8.jar -e 'require "io/console"; puts "we did it!!!!";'
io/console not supported; tty will not be manipulated
we did it!!!!
```
```
bash-3.2$ java -jar jruby-complete-9.1.7.0.jar -e 'require "io/console"; puts "we did it!!!!";'
TypeError: no implicit conversion of nil into String
   reopen at org/jruby/RubyIO.java:599
   <main> at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/io/console.rb:224
  require at org/jruby/RubyKernel.java:961
   (root) at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb:1
   <main> at -e:1
```

As for how I'm using it in my project?  Well I'm not actually using io/console directly in any of my own code but some of my gem dependencies do.  From the traceback it would appear that Bundler is the first such gem to trip this.

**Comment by headius** [comment]
I have had little luck finding an AIX machine I can test against, so we're going to have to do some back-and-forth here.

Run that again but pass `-Xbacktrace.style=full` to JRuby, so I can see the full Java trace.

**Comment by iconoclast** [comment]
No problem.  Happy for the help.

Well, -Xbacktrace.style=full just complained `jruby: invalid extended option backtrace.style=full (-X will list valid options)`
But I had a little better luck with -Djruby.backtrace.style=full

```
bash-3.2$ java -Djruby.backtrace.style=full -jar jruby-complete-9.1.7.0.jar -e 'require "io/console"; puts "we did it!!!!";' 
TypeError: no implicit conversion of nil into String
                  getStackTraceImpl at java/lang/Thread.java:-2
                      getStackTrace at java/lang/Thread.java:1114
                   getBacktraceData at org/jruby/runtime/backtrace/TraceType.java:246
                       getBacktrace at org/jruby/runtime/backtrace/TraceType.java:47
                   prepareBacktrace at org/jruby/RubyException.java:235
                           preRaise at org/jruby/exceptions/RaiseException.java:216
                           preRaise at org/jruby/exceptions/RaiseException.java:183
                             <init> at org/jruby/exceptions/RaiseException.java:111
                  newRaiseException at org/jruby/Ruby.java:4109
                       newTypeError at org/jruby/Ruby.java:3780
  handleImplicitlyUncoercibleObject at org/jruby/util/TypeConverter.java:400
            handleUncoercibleObject at org/jruby/util/TypeConverter.java:396
                    convertToType19 at org/jruby/util/TypeConverter.java:99
                    convertToType19 at org/jruby/util/TypeConverter.java:155
                      convertToType at org/jruby/util/TypeConverter.java:123
                    convertToString at org/jruby/RubyBasicObject.java:781
                           get_path at org/jruby/RubyFile.java:1398
                             reopen at org/jruby/RubyIO.java:599
                               call at org/jruby/RubyIO$INVOKER$i$0$1$reopen.gen:-1
                               call at org/jruby/internal/runtime/methods/JavaMethod.java:720
                               call at org/jruby/internal/runtime/methods/DynamicMethod.java:200
                       cacheAndCall at org/jruby/runtime/callsite/CachingCallSite.java:338
                               call at org/jruby/runtime/callsite/CachingCallSite.java:163
                        processCall at org/jruby/ir/interpreter/InterpreterEngine.java:315
                          interpret at org/jruby/ir/interpreter/StartupInterpreterEngine.java:73
                     INTERPRET_ROOT at org/jruby/ir/interpreter/Interpreter.java:112
                             <main> at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/io/console.rb:224
                            execute at org/jruby/ir/interpreter/Interpreter.java:99
                            execute at org/jruby/ir/interpreter/Interpreter.java:35
                            execute at org/jruby/ir/IRTranslator.java:42
                     runInterpreter at org/jruby/Ruby.java:835
                           loadFile at org/jruby/Ruby.java:2872
                               load at org/jruby/runtime/load/LibrarySearcher.java:243
                               load at org/jruby/runtime/load/LibrarySearcher.java:34
          tryLoadingLibraryOrScript at org/jruby/runtime/load/LoadService.java:887
                  smartLoadInternal at org/jruby/runtime/load/LoadService.java:527
                            require at org/jruby/runtime/load/LoadService.java:398
                      requireCommon at org/jruby/RubyKernel.java:968
                          require19 at org/jruby/RubyKernel.java:961
                               call at org/jruby/RubyKernel$INVOKER$s$1$0$require19.gen:-1
                               call at org/jruby/internal/runtime/methods/JavaMethod.java:383
                               call at org/jruby/internal/runtime/methods/AliasMethod.java:61
                               call at org/jruby/runtime/callsite/CachingCallSite.java:161
                        processCall at org/jruby/ir/interpreter/InterpreterEngine.java:315
                          interpret at org/jruby/ir/interpreter/StartupInterpreterEngine.java:73
                          interpret at org/jruby/ir/interpreter/InterpreterEngine.java:84
                   INTERPRET_METHOD at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:179
                             (root) at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb:1
                               call at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:165
                               call at org/jruby/internal/runtime/methods/DynamicMethod.java:200
                       cacheAndCall at org/jruby/runtime/callsite/CachingCallSite.java:338
                               call at org/jruby/runtime/callsite/CachingCallSite.java:163
               invokeOther1:require at -e:1
                             <main> at -e:1
                invokeWithArguments at java/lang/invoke/MethodHandle.java:463
                               load at org/jruby/ir/Compiler.java:90
                          runScript at org/jruby/Ruby.java:823
                        runNormally at org/jruby/Ruby.java:742
                        runNormally at org/jruby/Ruby.java:760
                        runFromMain at org/jruby/Ruby.java:573
                      doRunFromMain at org/jruby/Main.java:417
                        internalRun at org/jruby/Main.java:305
                                run at org/jruby/Main.java:232
                               main at org/jruby/Main.java:204
```

**Comment by DanHeidinga** [comment]
For access to AIX (and Linux on Power), check out the "IBM Power Development Cloud (PDP)": 

https://www-356.ibm.com/partnerworld/wps/servlet/ContentHandler/stg_com_sys_power-development-platform

I'm told this is a way to get access to Power machines for debugging / testing.  Caveat: I haven't tried it yet as my access request is pending.

**Comment by iconoclast** [comment]
@headius any other info you need from this end that may be of help?

**Comment by headius** [comment]
I think we need to update this bug and see if it's still working with recent fixes and rebuilds of the AIX support.

**Comment by headius** [comment]
May be fixed in 9.2 with updated jffi AIX support. Need someone to confirm whether it still exists. Bumped to 9.2.1 in case it's still broken.

**Comment by headius** [comment]
No confirmation and good chance this is fixed with updated AIX support. Optimistically closing.
